### ISSUE TITLE
[Feature Request] Revert part of changes from commit

### ISSUE THREAD

**Issue by sm-g** [ForkIssues/TrackerWin#148]
Found feature from SourceTree that I'm missing in Fork: revert part of commit ("reverse hunk / line" buttons).

**Comment by markjerz** [comment]
I think this already exists...

When you hover over a hunk of text in the diff you get "Stage" and "Discard" buttons at the top right of that hunk.

**Comment by DanPristupov** [comment]
@markjerz he meant the ability to revert a part of existing commit (in the All Commits view)

**Comment by ransagy** [comment]
Exactly what i came across today; Had a big complicated merge between two branches where i made some mistakes and hoped i can undo the specific hunks/lines i know were wrong from the commit/history view - but only had 'copy' as an option.

Instead i guess i'll have to undo/reset the entire file and then clean up the new commited files one by one.

**Comment by escamoteur** [comment]
Yep, that would be awesome. Or even more general being able to select a block on one branch/revision and merge into another branch/revision.

**Comment by DanPristupov** [comment]
We have this feature implemented but decided to not release because it causes lots of conflicts and is error prone.

**Comment by escamoteur** [comment]
Canceled forever or just till a better solution? If you mess up a big merge this could be really helpful

**Comment by DanPristupov** [comment]
It looks like a logical problem, not a technical one. You can only apply a patch if its context exactly matches the current one otherwise it will be a conflict. So, in reality only very recent commits can be reverted/copied and this is very confusing.

I think `Reset file to state at/before commit` + discard what is not needed is the best way to discard or copy changes in a particular file.

If we find how to make it work, we'll be happy to implement it of course.

**Comment by escamoteur** [comment]
reverting a certain marked block of changes should also be possible.
to allow patches like that you would need to mark source and dest areas. That would save a lot of time compared to how you have to do it now.
